Add a driver for the low side switch BD8LB600FS

benediktibk opened the pull request:
Add support to zephyr for the low side switch BD8LB600FS by Rohm.

benediktibk committed:
dts: bindings: gpio: add binding for BD8LB600FS

Add the binding for the driver of the chip BD8LB600FS.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk commented:
@henrikbrixandersen @cfriedt Looks like @mnkp hasn't been too active in the past few months, maybe one of you could take care of this pull request?

cfriedt commented:
@benediktibk - we're currently in a feature freeze for the v3.3 release.

I would suggest ensuring that checks pass before marking the PR ready for review. If the failure is not related to your change, you may need to rebase.

benediktibk commented:
Yes, I know about the feature freeze. I just wanted to use the time until then to get the PR ready to be merged.

The failing test is not related to my change:

`Assertion failed at WEST_TOPDIR/zephyr/tests/net/tcp/src/main.c:205: test_sem_take: (false is false)`

I'm gonna rebase the PR to trigger a new build.

cfriedt commented on drivers/gpio/gpio_bd8lb600fs.h:
```diff
@@ -0,0 +1,45 @@
+extern "C" {
+#endif
+
+/** Configuration data */
```
Please move the config and data structures to the `.c` file. This header can be removed.

benediktibk replied:
Okay for me. The only issue I have is that I am in general a lazy person and I basically just copied and modified the driver of MCP23S17 (https://github.com/zephyrproject-rtos/zephyr/blob/d56ed39e209d63f95593218f95fd04ea97d5bea6/drivers/gpio/gpio_mcp23s17.h). Therefore, I assume it is also wrong for the MCP23S17?

cfriedt replied:
I would say so, yea

benediktibk replied:
done

benediktibk replied:
Applied the same changes to the origin: #54814

cfriedt commented on drivers/gpio/gpio_bd8lb600fs.c:
```diff
@@ -0,0 +1,266 @@
+
+#define DT_DRV_COMPAT rohm_bd8lb600fs
+
+/**
```
I would suggest removing doxygen from this file.

benediktibk replied:
done

cfriedt commented on drivers/gpio/gpio_bd8lb600fs.c:
```diff
@@ -0,0 +1,266 @@
+#define WAIT_TIME_RESET_ACTIVE_IN_US	     1000
+#define WAIT_TIME_RESET_INACTIVE_TO_CS_IN_US 10
+
+/*!
```
I would suggest removing doxygen from this file.

benediktibk replied:
done

cfriedt commented on drivers/gpio/gpio_bd8lb600fs.c:
```diff
@@ -0,0 +1,266 @@
+	.pin_interrupt_configure = bd8lb600fs_pin_interrupt_configure,
+};
+
+/**
```
I would suggest removing doxygen from this file.

cfriedt commented on samples/drivers/gpio_bd8lb600fs/README.rst:
```diff
@@ -0,0 +1,50 @@
+.. _gpio_bd8lb600fs_sample:
```
Hmm... this is interesting. So normally, I would advise against a driver-implementation specific sample, but of course this is an interesting case where really any board with an exposed SPI port could be used to test this SPI GPIO expander.

Really, we do need to add entries for this driver to:
* `tests/drivers/build_all/gpio/`
* `tests/drivers/gpio/gpio_api_1pin/`
* `tests/drivers/gpio/gpio_basic_api/`

Since the peripheral is not hard-wired to the `nucleo_f429zi` board, it might make sense to include a `gpio_bd8lb600fs.overlay` and a `gpio_bd8lb600fs.conf` in the test directories, and to include additional `extra_configs` or `extra_args` as necessary to a new entry in `testcase.yaml`.

Still buildable with west as well, using the same `extra_configs` or `extra_args`.

Now - the fact that there is a state to restore (presumably this is w.r.t. power management), might need to be handled in a separate way, but lets get to that bridge after the next rev.

benediktibk replied:
I will try to figure this out and add the necessary entries. The reason for the sample was that I needed to test it anyway, hence I thought I could just submit my test project as well. It will probably make it easier for somebody else to use this driver as well, as there is a template on how it can be configured.

cfriedt replied:
Definitely would be helpful to others - no argument there.

The main issue is we have to try to keep samples to a minimum just because there are just so many unique devices and boards. 

At the same time, we want all GPIO devices to look and feel approximately the same.

With that, generic tests (and device-specific `.overlay` and `.conf`) should be enough to get people off the ground.

benediktibk replied:
I did not add this specific driver to gpio_basic_api, as it can only be used as an output. Otherwise I went ahead and tried to add this, hopefully correctly.

benediktibk committed:
drivers: gpio: add driver for BD8LB600FS

Add a driver for BD8LB600FS

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk committed:
tests: drivers: gpio: add BD8LB600FS

Add the BD8LB600FS to the build all tests,
as well as add another specific test for just
this chip which also demonstrates the
usage.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

cfriedt commented on dts/bindings/gpio/rohm,bd8lb600fs.yaml:
```diff
@@ -0,0 +1,31 @@
+  ngpios:
+    type: int
+    required: true
+    const: 8
```
I think this should not be const. It should be set at the device tree level and not at the binding level.

benediktibk replied:
I went ahead and checked the current state of the ngpios property of other GPIO controllers. Interestingely, there are a bunch of other bindings which also define this property as const:

- cypress,cy8c95xx-gpio-port
- microchip,mcp23s17
- nxp,pcal6408a
- nxp,pcf8574
- semtech,sx1509b
- ti,sn74hc595
- ti,tca9538

And then there is the case of these, which define it as enum, hence limited to a set of values:

- microchip,mcp23sxx
- microchip,mcp230xx

All these bindings have in common, that they are basically GPIO expanders. One is an 8 bit shift register, but for the sake of this argument I think this doesn't make a huge difference.

Going back to the description of ngpios in the GPIO controller I think it is correct ot define this property as const for an external chip:

> This property indicates the number of in-use slots of available slots
> for GPIOs. The typical example is something like this: the hardware
> register is 32 bits wide, but only 18 of the bits have a physical
> counterpart. The driver is generally written so that all 32 bits can
> be used, but the IP block is reused in a lot of designs, some using
> all 32 bits, some using 18 and some using 12. In this case, setting
> "ngpios = <18>;" informs the driver that only the first 18 GPIOs, at
> local offset 0 .. 17, are in use.  For cases in which there might be
> holes in the slot range, this value should be the max slot number-1.

Considering this it makes sense to keep this property as not const for all the MCU internal GPIO controllers, as in these cases it really will differ on the actual type of the MCU how many GPIOs the driver will have to support, as the driver will be reused for multiple MCUs from the same family. In this case, as this driver and binding are specific for exactly one chip which has a constant number of GPIOs I think I should stick consequently to how this property is defined for other GPIO expanders.

henrikbrixandersen replied:
Please see my reply to a similar concern raised by @cfriedt here: https://github.com/zephyrproject-rtos/zephyr/pull/55093#discussion_r1118462208

henrikbrixandersen replied:
> Considering this it makes sense to keep this property as not const for all the MCU internal GPIO controllers, as in these cases it really will differ on the actual type of the MCU how many GPIOs the driver will have to support, as the driver will be reused for multiple MCUs from the same family. In this case, as this driver and binding are specific for exactly one chip which has a constant number of GPIOs I think I should stick consequently to how this property is defined for other GPIO expanders.

Agreed.

henrikbrixandersen commented on tests/drivers/gpio/gpio_api_1pin/testcase.yaml:
```diff
@@ -1,8 +1,14 @@
     platform_exclude: mps2_an385 mps2_an521 neorv32
     filter: dt_enabled_alias_with_parent_compat("led0", "gpio-leds")
+
+  peripheral.gpio.bd8lb600fs:
+    platform_allow: native_posix
+    extra_args: DTC_OVERLAY_FILE="boards/gpio_bd8lb600fs.overlay"
```
How is this supposed to work on native posix?

benediktibk replied:
Good point, I am actually uncertain what I should put there, Because there is no board available with this hardware soldered onto it, I just have my test setup which I built for myself.

henrikbrixandersen replied:
You could add a generic shield definition using this chip. That what I have done in the past.

benediktibk replied:
Could you maybe point me to an example as a starting point?

henrikbrixandersen replied:
One possible source of inspiration could be https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/shields/ssd1306

benediktibk replied:
Thanks, I figured out you used this shield for the display driver sample.

So what I could do would be to add this shield and use it for this testcase. But I assume nobody will every specifically add a test with hardware in the loop for exactly this setup. Therefore, I don't  think it is worth it to add this test. I will remove it again.
